### ISSUE TITLE
50% of deploys returns with error "Release directory not empty"

### ISSUE THREAD

**Issue by samplenull** [deployphp/deployer#1287]
Issue Type: Bug
Deployer Version: v5.1.3
Local Machine OS: Ubuntu 16.04.2 LTS
Remote Machine OS: Ubuntu 16.04.2 LTS

### Description 
Every second deploy breaks with error about directory already exists. I've tried to clean up everything and start releases again but eventually came to the same error.

### Content of `deploy.php`
```php
// Project deploy settings
require __DIR__ . '/vendor/deployer/deployer/recipe/symfony.php';
//....
add('shared_files', [
    'app/config/parameters.yml',
]);
set('writable_use_sudo', true);
set('writable_mode', 'chmod');
set('writable_chmod_mode', '0775'); // For chmod mode

add('shared_dirs', [
    'var/logs',
    'vendor',
    'web/assets/vendor',
    'web/resource',
]);

add('writable_dirs', [
    'var/cache',
    'var/logs',
    'web/resource',
]);

set('cleanup_use_sudo', true);

//....

/**
 * Main task
 */
task('deploy', [
    'deploy:prepare',
    'deploy:lock',
    'deploy:writable',
    'deploy:release',
    'deploy:update_code',
    'deploy:create_cache_dir',
    'deploy:shared',
    'deploy:assets',
    'deploy:vendors',
    'deploy:assetic:dump',
    'deploy:writable',
    'deploy:cache:warmup',
    'deploy:api:cache:clear',
    'deploy:symlink',
    'deploy:unlock',
    'deploy:frontend:update',
    'deploy:writable',
    'cleanup',
])->desc('Deploy your project');
//...
```

### Output log
*With enabled option for verbose output `-vvv`.*

```
+ echo ========================================== Initializing deploy ==========================================
+ cd /var/deployer/deployer-config/project
+ export DEPLOY_PATH=/var/www/project-backend
+ php vendor/bin/dep deploy dev -vvv
PHP Notice:  Only variables should be passed by reference in /var/deployer/deployer-config/project/deploy.php on line 28
➤ Executing task deploy:prepare
[project-dev] > echo $0
[project-dev] < ssh multiplexing initialization
[project-dev] < Welcome to Ubuntu 16.04.2 LTS (GNU/Linux 2.6.32-042stab123.2 x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

[project-dev] < bash
[project-dev] > if [ ! -d /var/www/project-backend ]; then mkdir -p /var/www/project-backend; fi
[project-dev] > if [ ! -L /var/www/project-backend/current ] && [ -d /var/www/project-backend/current ]; then echo true; fi
[project-dev] > cd /var/www/project-backend && if [ ! -d .dep ]; then mkdir .dep; fi
[project-dev] > cd /var/www/project-backend && if [ ! -d releases ]; then mkdir releases; fi
[project-dev] > cd /var/www/project-backend && if [ ! -d shared ]; then mkdir shared; fi
• done on [project-dev]
✔ Ok [496ms]
➤ Executing task deploy:lock
• done on [project-dev]
✔ Ok [0ms]
➤ Executing task deploy:writable
[project-dev] > if [ -h /var/www/project-backend/release ]; then echo 'true'; fi
[project-dev] > readlink /var/www/project-backend/current
[project-dev] < releases/38
[project-dev] > cd /var/www/project-backend/releases/38 && (mkdir -p app/cache app/logs var/cache var/logs web/resource)
[project-dev] > cd /var/www/project-backend/releases/38 && (sudo chmod -R 0775 app/cache app/logs var/cache var/logs web/resource)
• done on [project-dev]
✔ Ok [369ms]
➤ Executing task deploy:release
[project-dev] > cd /var/www/project-backend && (if [ -h release ]; then echo 'true'; fi)
[project-dev] > cd /var/www/project-backend && ([ -d releases ] && [ "$(ls -A releases)" ] && echo "true" || echo "false")
[project-dev] < true
[project-dev] > cd /var/www/project-backend && (cd releases && ls -t -1 -d */)
[project-dev] < 38/
[project-dev] < 37/
[project-dev] < 36/
[project-dev] < 35/
[project-dev] < 34/
[project-dev] < 3/
[project-dev] < 2/
[project-dev] < 1/
[project-dev] < 4/
[project-dev] > cd /var/www/project-backend && (if [ -f .dep/releases ]; then echo "true"; fi)
[project-dev] < true
[project-dev] > cd /var/www/project-backend && (tail -n 15 .dep/releases)
[project-dev] < 20170627143313,29
[project-dev] < 20170628105946,30
[project-dev] < 20170628124035,31
[project-dev] < 20170629151651,32
[project-dev] < 20170630132639,33
[project-dev] < 20170703110822,34
[project-dev] < 20170703112129,34
[project-dev] < 20170703173741,35
[project-dev] < 20170703175421,35
[project-dev] < 20170704111141,36
[project-dev] < 20170704111444,36
[project-dev] < 20170704112435,37
[project-dev] < 20170704112450,37
[project-dev] < 20170704124653,38
[project-dev] < 20170704124810,38
[project-dev] > cd /var/www/project-backend && (if [ -d /var/www/project-backend/releases/39 ]; then echo 'true'; fi)
[project-dev] > cd /var/www/project-backend && (date +"%Y%m%d%H%M%S")
[project-dev] < 20170704125757
[project-dev] > cd /var/www/project-backend && (echo '20170704125757,39' >> .dep/releases)
[project-dev] > cd /var/www/project-backend && (mkdir /var/www/project-backend/releases/39)
[project-dev] > cd /var/www/project-backend && (if [[ $(man ln 2>&1 || ln -h 2>&1 || ln --help 2>&1) =~ '--relative' ]]; then echo 'true'; fi)
[project-dev] < true
[project-dev] > cd /var/www/project-backend && (ln -nfs --relative /var/www/project-backend/releases/39 /var/www/project-backend/release)
• done on [project-dev]
✔ Ok [409ms]
➤ Executing task deploy:update_code
[project-dev] > if hash command 2>/dev/null; then echo 'true'; fi
[project-dev] < true
[project-dev] > command -v 'git'
[project-dev] < /usr/bin/git
[project-dev] > /usr/bin/git version
[project-dev] < git version 2.7.4
[project-dev] > /usr/bin/git clone -b master --recursive -q --reference /var/www/project-backend/releases/38 --dissociate git@bitbucket.org:uts-group/project-backend.git  /var/www/project-backend/releases/38 2>&1
[project-dev] < fatal: destination path '/var/www/project-backend/releases/38' already exists and is not an empty directory.
[project-dev] > /usr/bin/git clone -b master --recursive -q git@bitbucket.org:uts-group/project-backend.git /var/www/project-backend/releases/38 2>&1
[project-dev] < fatal: destination path '/var/www/project-backend/releases/38' already exists and is not an empty directory.
➤ Executing task deploy:failed
• done on [project-dev]
✔ Ok [0ms]

                                                                               
  [Deployer\Exception\RuntimeException (128)]                                  
  The command "/usr/bin/git clone -b master --recursive -q git@bitbucket.org:  
  uts-group/project-backend.git /var/www/project-backend/releases/38 2>&1" failed.   
  Exit Code: 128 (Invalid exit argument)                                       
  Host Name: project-dev                                                          
  ================        

```

**Comment by antonmedv** [comment]
Try delete this line:
```diff
task('deploy', [
    'deploy:prepare',
    'deploy:lock',
-   'deploy:writable',
    'deploy:release',
    'deploy:update_code',
    'deploy:create_cache_dir',
    'deploy:shared',
    'deploy:assets',
    'deploy:vendors',
    'deploy:assetic:dump',
    'deploy:writable',
    'deploy:cache:warmup',
    'deploy:api:cache:clear',
    'deploy:symlink',
    'deploy:unlock',
    'deploy:frontend:update',
    'deploy:writable',
    'cleanup',
])->desc('Deploy your project');
```

It causes precache release to current

**Comment by samplenull** [comment]
I added deploy:writable because I had some issues with cache folders removal, i guess the only one solution is deploy via www-data user?

**Comment by antonmedv** [comment]
Use proper user configuration.

**Comment by samplenull** [comment]
> It causes precache release to current
@antonmedv Thanks, didn't know that, removed 2 first "writables" and no errors since then

**Comment by antonmedv** [comment]
I think, I'll add check in `deploy:release`.

**Comment by Ikstar** [comment]
I am having a similar issue. It tries to deploy to the the previous release

````
 ➤ Executing task deploy:prepare
[qa] > echo $0
[qa] > if [ ! -d /var/www/sites/platform ]; then mkdir -p /var/www/sites/platform; fi
[qa] > if [ ! -L /var/www/sites/platform/current ] && [ -d /var/www/sites/platform/current ]; then echo true; fi
[qa] > cd /var/www/sites/platform && if [ ! -d .dep ]; then mkdir .dep; fi
[qa] > cd /var/www/sites/platform && if [ ! -d releases ]; then mkdir releases; fi
[qa] > cd /var/www/sites/platform && if [ ! -d shared ]; then mkdir shared; fi
• done on [qa]
✔ Ok [317ms]
➤ Executing task deploy:lock
[qa] > if [ -f /var/www/sites/platform/.dep/deploy.lock ]; then echo 'true'; fi
[qa] > touch /var/www/sites/platform/.dep/deploy.lock
• done on [qa]
✔ Ok [108ms]
➤ Executing task deploy:release
[qa] > cd /var/www/sites/platform && (if [ -h release ]; then echo 'true'; fi)
[qa] > cd /var/www/sites/platform && ([ -d releases ] && [ "$(ls -A releases)" ] && echo "true" || echo "false")
[qa] > cd /var/www/sites/platform && (cd releases && ls -t -1 -d */)
[qa] > cd /var/www/sites/platform && (if [ -f .dep/releases ]; then echo "true"; fi)
[qa] > cd /var/www/sites/platform && (tail -n 15 .dep/releases)
[qa] > cd /var/www/sites/platform && (if [ -d /var/www/sites/platform/releases/9 ]; then echo 'true'; fi)
[qa] > cd /var/www/sites/platform && (date +"%Y%m%d%H%M%S")
[qa] > cd /var/www/sites/platform && (echo '20170719160910,9' >> .dep/releases)
[qa] > cd /var/www/sites/platform && (mkdir /var/www/sites/platform/releases/9)
[qa] > cd /var/www/sites/platform && (if [[ $(man ln 2>&1 || ln -h 2>&1 || ln --help 2>&1) =~ '--relative' ]]; then echo 'true'; fi)
[qa] > cd /var/www/sites/platform && (ln -nfs --relative /var/www/sites/platform/releases/9 /var/www/sites/platform/release)
• done on [qa]
✔ Ok [606ms]
➤ Executing task deploy:update_code
[qa] > if hash command 2>/dev/null; then echo 'true'; fi
[qa] > command -v 'git'
[qa] > /usr/bin/git version
[qa] > /usr/bin/git clone -b master --depth 1 --recursive -q git@github.com:myrepoplace/myrepoplace.git /var/www/sites/platform/releases/8 2>&1
➤ Executing task deploy:failed
• done on [qa]
✔ Ok [0ms]
➤ Executing task deploy:unlock
[qa] > rm -f /var/www/sites/platform/.dep/deploy.lock
• done on [qa]
✔ Ok [82ms]


  [Deployer\Exception\RuntimeException (128)]
  The command "/usr/bin/git clone -b master --depth 1 --recursive -q git@github.com:myrepoplace/myrepoplace.git /
  var/www/sites/platform/releases/8 2>&1" failed.
  Exit Code: 128 (Invalid exit argument)
  Host Name: qa
  ================

```

It seems to grab the releases/8 folder instead of the newly created releases/9
This is happening with the stock laravel deployment recipe.

**Comment by antonmedv** [comment]
Should be fixed in #1904
